### PR TITLE
fix: do not re-use readstreams when retry an upload of the test replay recording

### LINKED ISSUE #29227: Protocol: [Upload] Do not re-use old fs ReadStream on fetch retry
### Problem

Once a ReadStream is opened by a destination stream, it cannot be used by any other stream.

Currently, the Test Replay ReadStream is re-used between retrying `fetch` when an error occurs. If `fetch` opens this readstream before encountering and throwing an error, the readstream should be closed and unable to be re-opened. This may be (but is not yet proven to be) causing erratic error messages when retrying Test Replay artifact uploads.

### Possible Solution
before `fetch` is retried, the fs ReadStream for the source archive should be destroyed and re-created.

### PR DESCRIPTION
**Pull request by cacieprins**
- Closes https://github.com/cypress-io/cypress/issues/29227

### Additional details
If `node-fetch` began reading from the Test Replay recording archive for uploading, and encountered a retryable error, it would attempt to re-use the same ReadStream on subsequent retries. These retries would fail, because ReadStreams may only be opened for reading once.

This PR refactors the previous `upload_stream` logic:
* A generic `asyncRetry` function is used instead of `fetch-retry-ts`. This function can be subsequently used in other refactors (e.g., reconnection in `CriClient`).
* A `putFetch` function is used to determine Http vs. Network vs. response Parsing errors, for more streamlined error handling. This pattern can be subsequently re-used and reduced via DRY for other HTTP methods. Especially useful is the ability to define how the response body should be parsed (other options can be added as needed, like `Blob` or `ArrayBuffer`), and what the type of the response should be. This will aid in any future refactors of other API methods.
* The logic for creating the artifact ReadStream and StreamMonitor (for stall detection) is moved to `put_protocol_artifact.ts`, which is wrapped in `asyncRetry` and leverages `putFetch`, all of which simplifies error handling and ensures both ReadStream and StreamMonitor are instantiated fresh for each upload attempt.

A subsequent PR will be opened to remove the now-unreferenced `fetch-retry-ts` package.

### Steps to test
Run the `record_spec` system tests; run a test recording against staging; (tbd) use Charles proxy or similar to simulate initial failures and subsequent successes on upload

### How has the user experience changed?
Retry attempts to upload that previously failed due to Stream issues may retry properly now. This will not resolve issues that cause the initial (or subsequent) upload attempt to fail caused by the upstream server or local network conditions, it simply ensures that we are retrying properly.

### PR Tasks

- [X] Have tests been added/updated?
- [NA] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [NA] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by jennifer-shehane** [comment]
@cacieprins Looks like the snapshots are still being weird. 
Screenshot 2024-06-26 at 10 22 47 AM

**Comment by cacieprins** [comment]
I'm investigating. The snapshot that's failing is asserting shape of a POST to `instances/:id/results`; the contents of which should not have changed from this PR. The snapshot itself was modified, but only by some changes in whitespace - not nearly to the degree that it's failing is mismatch. I'm not able to get the same snapshot results locally, so far.

File: packages/server/lib/cloud/api/put_fetch.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+  try {
+    const response = await (crossFetch as typeof crossFetch)(input, {
+      ...(init || {}),
+      method: 'PUT',
+      // cross-fetch thinks this is in the browser, so declares
+      // types based on that rather than on node-fetch which it
+      // actually uses under the hood. node-fetch supports `agent`.
+      // @ts-expect-error
+      agent,
```
Comment (cacieprins) [inline packages/server/lib/cloud/api/put_fetch.ts]: using node-fetch directly would probably remove the need for this @ts-expect-error

File: packages/server/lib/cloud/api/put_protocol_artifact.ts [hunk 2]
```diff
@@ -13,23 +16,44 @@ const debug = Debug('cypress:server:cloud:api:protocol-artifact')
...
+      parse: ParseKinds.TEXT,
+      headers: {
+        'content-length': String(size),
+        'content-type': 'application/x-tar',
+        'accept': 'application/json',
+      },
+      // ts thinks this is a web fetch, which only expects ReadableStreams.
+      // But, this is a node fetch, which supports ReadStreams.
+      // @ts-expect-error
+      body: activityMonitor.monitor(fileStream),
```
Comment (cacieprins) [inline packages/server/lib/cloud/api/put_protocol_artifact.ts:42]: Using `node-fetch` directly would likely remove the need for this `ts-expect-error`

File: packages/server/test/unit/cloud/api/put_protocol_artifact_spec.ts [hunk 3]
```diff
@@ -146,37 +170,31 @@ describe('putProtocolArtifact', () => {
...
-            activityMonitor: mockStreamMonitor,
-          },
-        ).resolves()
+      it('resolves', async () => {
+        putFetchStub.resolves()
 
-        return expect(putProtocolArtifact(filePath, maxFileSize, destinationUrl)).to.be.fulfilled
+        await putProtocolArtifact(filePath, maxFileSize, destinationUrl)
+
+        expect(true).to.be.true
```
Comment (AtofStryker) [inline packages/server/test/unit/cloud/api/put_protocol_artifact_spec.ts]: do we have something to assert here in this test?

File: packages/server/lib/cloud/network/put_fetch.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  TReturn extends any
+> (input: RequestInfo | URL, options: PutOptions = { parse: 'json' }): Promise<TReturn> {
+  const {
+    parse,
+    ...init
+  } = options
+
+  debug('Initiating PUT %s', input)
+  try {
+    const response = await (crossFetch as typeof crossFetch)(input, {
```
Comment (ryanthemanuel) [inline packages/server/lib/cloud/network/put_fetch.ts]: Why is:

```
crossFetch as typeof crossFetch
```

necessary?

Comment (cacieprins) [reply packages/server/lib/cloud/network/put_fetch.ts]: It isn't - this was vestigial from trying to resolve the typescript error that gets ignored a few lines below. removed in https://github.com/cypress-io/cypress/pull/29745/commits/e0581d7b46a288af1f336c04f6ca76ee61e8ced8

File: packages/server/test/unit/cloud/api/put_protocol_artifact_spec.ts (hunk 3 above)
Comment (cacieprins) [reply packages/server/test/unit/cloud/api/put_protocol_artifact_spec.ts]: this is basically the same as chai as promised' expect().to.be.fulfilled - i changed it back to be more clear. I'd unwrapped it for debugging purposes - thanks for the catch! addressed in https://github.com/cypress-io/cypress/pull/29745/commits/401302054063ee68d9a2c0990c6899f1b82f295f

File: cli/CHANGELOG.md [hunk 5]
```diff
@@ -1,4 +1,12 @@
 <!-- See the ../guides/writing-the-cypress-changelog.md for details on writing the changelog. -->
+## 13.13.1
+
+_Release 7/16/2024 (PENDING)_
```
Comment (cacieprins) [inline cli/CHANGELOG.md]:
```suggestion
_Released 7/16/2024 (PENDING)_
```

**Review by ryanthemanuel — APPROVED** [review]

File: system-tests/__snapshots__/record_spec.js [hunk 6]
```diff
@@ -3049,6 +3049,14 @@ exports['e2e record capture-protocol enabled protocol runtime errors db size too
   - Screenshot - Done Uploading 1 kB in Xm, Ys ZZ.ZZms 1/2 /XXX/XXX/XXX/cypress/screenshots/record_pass.cy.js/yay it passes.png
   - Test Replay - Failed Uploading after Xm, Ys ZZ.ZZms 2/2 - Spec recording too large: artifact is 1024 bytes, limit is 200 bytes
 
+Warning: We encountered an HTTP error while uploading the Test Replay recording for this spec.
```
Comment (cacieprins) [inline system-tests/__snapshots__/record_spec.js]: Missed this delta - it should not be displaying this error here

File: packages/server/test/unit/cloud/network/is_retryable_error_spec.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+chai.use(sinonChai)
+
+describe('isRetryableError', () => {
+  const url = 'http://some/url'
+
+  it('returns true with a NetworkError', () => {
+    expect(isRetryableError(new NetworkError(new Error(), url))).to.be.true
+  })
+
+  it('returns true with retriable http errors', () => {
```
Comment (mschile) [inline packages/server/test/unit/cloud/network/is_retryable_error_spec.ts]: For consistency:
```suggestion
  it('returns true with retryable http errors', () => {
```

File: packages/server/test/unit/cloud/network/is_retryable_error_spec.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    [408, 429, 502, 503, 504].forEach((status) => {
+      const err = new HttpError('some error', url, status, 'status text', sinon.createStubInstance(Response))
+
+      expect(isRetryableError(err)).to.be.true
+    })
+  })
+
+  it('returns false with non-retriable http errors', () => {
```
Comment (mschile) [inline packages/server/test/unit/cloud/network/is_retryable_error_spec.ts]: For consistency:
```suggestion
  it('returns false with non-retryable http errors', () => {
```

File: system-tests/test/record_spec.js [hunk 9]
```diff
@@ -2689,7 +2689,7 @@ describe('capture-protocol api errors', () => {
 
         expect(artifactReport?.protocol).to.exist()
         expect(artifactReport?.protocol?.error).to.equal(
-          'Failed to upload Test Replay: 500 Internal Server Error (http://localhost:1234/capture-protocol/upload/?x-amz-credential=XXXXXXXX&x-amz-signature=XXXXXXXXXXXXX)',
+          'Failed to upload Test Replay after 1 attempts. Errors: 500 Internal Server Error (http://localhost:1234/capture-protocol/upload/?x-amz-credential=XXXXXXXX&x-amz-signature=XXXXXXXXXXXXX)',
```
Comment (mschile) [inline system-tests/test/record_spec.js]: This should be singular, `attempt`:

```suggestion
          'Failed to upload Test Replay after 1 attempt. Errors: 500 Internal Server Error (http://localhost:1234/capture-protocol/upload/?x-amz-credential=XXXXXXXX&x-amz-signature=XXXXXXXXXXXXX)',
```

File: system-tests/__snapshots__/record_spec.js [hunk 10]
```diff
@@ -3049,6 +3049,14 @@ exports['e2e record capture-protocol enabled protocol runtime errors db size too
...
+
+These results will not display Test Replay recordings.
+
+This error will not affect or change the exit code.
+
+undefined responded with HTTP undefined: undefined
```
Comment (mschile) [inline system-tests/__snapshots__/record_spec.js]: Should there be actual values here instead of `undefined`?

**Review by mschile — CHANGES_REQUESTED** [review]
Looks good overall, just added a few suggestions/questions.

File: system-tests/__snapshots__/record_spec.js (hunk 10 above)
Comment (cacieprins) [reply system-tests/__snapshots__/record_spec.js]: This style of error message shouldn't be displayed at all for this issue - fixed in https://github.com/cypress-io/cypress/pull/29745/commits/366f1d94a0a9e8c3b00105350410a6686bd1da0c

File: system-tests/test/record_spec.js (hunk 9 above)
Comment (cacieprins) [reply system-tests/test/record_spec.js]: fixed in https://github.com/cypress-io/cypress/pull/29745/commits/366f1d94a0a9e8c3b00105350410a6686bd1da0c

**Review by mschile — APPROVED** [review]

**Pull request merged**